File: src/main/java/com/kingdom/util/specialaction/FanSpecialActionHandler.java
package com.kingdom.util.specialaction;

import com.kingdom.model.*;

import java.util.HashSet;
import java.util.Set;

public class FanSpecialActionHandler {

    public static void handleSpecialAction(Game game, Card card) {
        Player player = game.getCurrentPlayer();

        if (card.getName().equals("Archivist")) {
            CardAction cardAction = new CardAction(CardAction.TYPE_CHOICES);
            cardAction.setDeck(Card.DECK_FAN);
            cardAction.setCardName(card.getName());
            cardAction.setInstructions("Choose one: Draw until you have 6 cards in hand; or +$1 and discard 1 or more cards.");
            cardAction.getChoices().add(new CardActionChoice("Draw", "draw"));
            cardAction.getChoices().add(new CardActionChoice("+$1 and discard", "discard"));
            game.setPlayerCardAction(player, cardAction);
        } else if (card.getName().equals("Museum")) {
            Set<Card> cards = new HashSet<Card>();
            for (Card c : player.getHand()) {
                if (!player.getMuseumCards().contains(c)) {
                    cards.add(c);
                }
            }
            if (!cards.isEmpty()) {
                CardAction museumCardAction = new CardAction(CardAction.TYPE_CHOOSE_UP_TO);
                museumCardAction.setDeck(Card.DECK_FAN);
                museumCardAction.setCardName(card.getName());
                museumCardAction.getCards().addAll(cards);
                museumCardAction.setNumCards(1);
                museumCardAction.setInstructions("Select a card to be added to your Museum mat and then click Done, or just click Done if you don't want to add a card.");
                museumCardAction.setButtonValue("Done");
                game.setPlayerCardAction(player, museumCardAction);
            } else if (player.getMuseumCards().size() >= 4) {
                CardAction museumCardAction = new CardAction(CardAction.TYPE_YES_NO);
                museumCardAction.setDeck(Card.DECK_FAN);
                museumCardAction.setCardName("Museum Trash Cards");
                museumCardAction.setInstructions("Do you want to trash 4 cards from your Museum mat to gain a Prize and a Duchy?");
                game.setPlayerCardAction(player, museumCardAction);
            }
        }
    }
}

File: src/main/kotlin/com/kingdom/util/specialaction/FanSpecialActionHandler.kt
package com.kingdom.util.specialaction

import com.kingdom.model.*
import com.kingdom.model.cards.Card
import com.kingdom.model.cards.Deck

object FanSpecialActionHandler {

    fun handleSpecialAction(game: Game, card: Card) {
        val player = game.currentPlayer

        when (card.name) {
            "Archivist" -> {
                val cardAction = CardAction(CardAction.TYPE_CHOICES)
                cardAction.deck = Deck.Fan
                cardAction.cardName = card.name
                cardAction.instructions = "Choose one: Draw until you have 6 cards in hand; or +$1 and discard 1 or more cards."
                cardAction.choices.add(CardActionChoice("Draw", "draw"))
                cardAction.choices.add(CardActionChoice("+$1 and discard", "discard"))
                game.setPlayerCardAction(player!!, cardAction)
            }
            "Museum" -> {
                val cards = player!!.hand
                        .filterNot { player.museumCards.contains(it) }
                        .toSet()
                if (!cards.isEmpty()) {
                    val museumCardAction = CardAction(CardAction.TYPE_CHOOSE_UP_TO)
                    museumCardAction.deck = Deck.Fan
                    museumCardAction.cardName = card.name
                    museumCardAction.cards.addAll(cards)
                    museumCardAction.numCards = 1
                    museumCardAction.instructions = "Select a card to be added to your Museum mat and then click Done, or just click Done if you don't want to add a card."
                    museumCardAction.buttonValue = "Done"
                    game.setPlayerCardAction(player, museumCardAction)
                } else if (player.museumCards.size >= 4) {
                    val museumCardAction = CardAction(CardAction.TYPE_YES_NO)
                    museumCardAction.deck = Deck.Fan
                    museumCardAction.cardName = "Museum Trash Cards"
                    museumCardAction.instructions = "Do you want to trash 4 cards from your Museum mat to gain a Prize and a Duchy?"
                    game.setPlayerCardAction(player, museumCardAction)
                }
            }
        }
    }
}
